### ISSUE TITLE
Hugo performance issue with recursively rendered nav

### ISSUE THREAD

**Issue by gumaerc** [mitodl/hugo-course-publisher#89]
If you look around for the best way to render a nav in Hugo with all its child elements, most of what you find recommends using a recursive template call with partials.  Here's an example:

https://gist.github.com/davidejones/4117fe17f96dac72346080734c44a7f5

We've implemented this in hugo-course-publisher and while it works, it has increased the site render time by nearly a factor of 10.  What used to take about 1 second now takes nearly 10 seconds.

With recursive nav enabled:
![image](https://user-images.githubusercontent.com/12089658/75559928-53cbd580-5a12-11ea-9611-2a663eeeee58.png)

With recursive nav disabled:
![image](https://user-images.githubusercontent.com/12089658/75560352-fc7a3500-5a12-11ea-8e5f-5ea054d20c7e.png)

The most concerning thing is the amount of calls to the `nav_item.html` partial.  This partial makes calls to both `IsMenuCurrent` and `HasMenuCurrent` for every nav item recursively.  In theory, this means that the lowest tier menu items have these called on them once for every tier of menus above them and once for themselves.  I'm unsure how to better optimize it at this point, and there doesn't seem to be a better way to render a non-deterministic amount of menu tiers.

The source of our recursive nav item partial is available here:
https://github.com/mitodl/hugo-course-publisher/blob/master/site/layouts/partials/nav_item.html

**Comment by bep** [comment]
Hey,

If that partial needs to be rendered 573488 times, that is taking some time.  I don't know your menu structure to understand if that number is correct. It takes 3,5 second on my fairly beefy MacBook.

The menu feature in Hugo isn't particularly fast if you want it to go faster, I currently think you make the menu navigation less page specific -- so it can somehow be cached with `partialCached`.

**Comment by gumaerc** [comment]
> Hey,
> 
> If that partial needs to be rendered 573488 times, that is taking some time. I don't know your menu structure to understand if that number is correct. It takes 3,5 second on my fairly beefy MacBook.
> 
> The menu feature in Hugo isn't particularly fast if you want it to go faster, I currently think you make the menu navigation less page specific -- so it can somehow be cached with `partialCached`.

Hey Bjørn,

Thanks for checking this out!  The site we are working on has a fairly extensive menu structure and items can be added to it dynamically.  I think using `partialCached` would save us significant time.  We're using bootstrap 4 collapse elements to tier the UI, and there is logic in the partial that puts CSS classes on collapse elements that need to be expanded by default on any given page as well as the active menu item that needs to be highlighted.  The pure CSS solution is nice, but if we handle this in Javascript on page load then the menu for each course could probably be cached and save a lot of time.

**Comment by bep** [comment]
I managed to delete my own comment, sorry about that. So, In general I think this issue is about lots of template rendering and not so much the speed of the menu system itself (e.g. `.HasMenuCurrent`).  You can probably save some by doing the recursiveness in an inline template (`define ...`), but probably not worth it.

**Comment by bep** [comment]
An added note. There is an issue with the `--templateMetricsHints` that I have now fixed in a PR (soon in a Hugo version). Running it on this project:

```bash
    cache     cumulative       average       maximum
  potential       duration      duration      duration  count  template
      -----     ----------      --------      --------  -----  --------
         34 8m44.640753859s     914.824µs   77.609449ms  573488  partials/nav_item.html
          4 5m30.075227801s  122.069241ms   242.09755ms   2704  partials/nav.html
          0 5m9.018112485s  264.118044ms  466.006965ms   1170  courses/single.html
          5 2m47.458551679s  123.859875ms  238.142768ms   1352  partials/desktop_nav.html
          5 2m47.227702556s  123.689129ms  246.303643ms   1352  partials/mobile_nav.html
          0  42.21955912s  231.975599ms  471.071852ms    182  courses/section.html
         24  5.756175098s    2.110808ms   13.314245ms   2727  partials/course_info.html
         29  3.782010292s    2.797344ms   13.521198ms   1352  partials/mobile_course_info.html
         15  2.090251289s      1.5728ms   17.130704ms   1329  partials/course_section.html
          0  219.684731ms       25.86µs    8.988655ms   8495  _internal/shortcodes/ref.html
         42   76.717517ms       9.377µs    1.526109ms   8181  partials/partial_collapse_list.html
          0   17.734578ms   17.734578ms   17.734578ms      1  _internal/_default/sitemap.xml
         56   16.321631ms      12.072µs     172.968µs   1352  partials/course_instructor_number.html
          0    1.772721ms         248ns      43.288µs   7146  shortcodes/br.html
          0    1.276771ms         943ns      58.914µs   1353  partials/header.html
        100    1.241423ms         918ns      62.761µs   1352  partials/mobile_nav_toggle.html
          0     834.743µs         616ns      63.213µs   1354  partials/footer.html
          0     557.653µs     557.653µs     557.653µs      1  courseindex/section.html
         72     419.133µs      18.223µs      94.028µs     23  partials/course_image.html
          0      82.896µs      82.896µs      82.896µs      1  404.html
          0      79.878µs      79.878µs      79.878µs      1  index.html

```

The first column: 100 means all of the rendered partials rendered the same, so it can (probably) be safely cached.

**Comment by gumaerc** [comment]
@bep Awesome, thank you for all your help!  This will be a great optimization tool to have.  Part of our issue originally was that we started this project based on this template: https://github.com/netlify-templates/victor-hugo.  The webpack setup in the template uses webpack-dev-server to host the site over hugo server.  Saving anything in our templates or CSS would cause webpack to rebuild, which then would trigger Hugo to rebuild in a cascade that would cause your browser tab to thrash from constant auto-refresh.  On master we are now using the Hugo development server and it is much better so this is less of an immediate issue.  I think we'd still like to get that build time down in the future, and caching will be the way to do it.  I think the Javascript solution I posted above will work best, because the main reason navs cannot be cached easily right now is the fact that we are injecting CSS into the template during rendering for each page that determines the state of expandables in the nav and highlighting the selected item.  If we didn't do this, the nav for each course would be the same on every page as far as the template engine is concerned.  We could then use a small JS script on page load to determine the selected nav item based on the current page.
